Adding Python samples to the Stateful Processing post.

pabloem opened the pull request:
Also add a couple more pydocs to python transforms.

r: @kennknowles 
cc: @mxm

pabloem commented:
Run Website PreCommit

mxm commented:
Thanks! Looks good. Have you tested this with an actual project? Just asking because we do not have checks for the doc code.

CC @robertwb

pabloem commented:
The indexing dofn is run as part of the unit tests for user state. The rest of the code I haven't ran... it's just a translation from the Java code to Python code (using BagState, as Python doesn't have a plain ValueState).

robertwb commented on website/src/_posts/2017-02-13-stateful-processing.md:
````diff
@@ -501,29 +543,60 @@ new DoFn<KV<UserId, Event>, KV<UserId, Prediction>>() {
 ```py
 # State and timers are not yet supported in Beam's Python SDK.
 # Watch this space!
````
You can remove these comment lines now.

pabloem replied:
Done.

robertwb commented on website/src/_posts/2017-02-13-stateful-processing.md:
```diff
@@ -501,29 +543,60 @@ new DoFn<KV<UserId, Event>, KV<UserId, Prediction>>() {
+            previous_prediction, new_prediction)):
+        previous_pred_state.clear()
+        previous_pred_state.add(new_prediction)
+        yield (user, new_prediction)
```
2 space indentation.

pabloem replied:
Done.

mxm commented:
On this note, could we add `ValueState` to the Python SDK. Perhaps just as a simple wrapper around ListState?

robertwb commented on website/src/_posts/2017-02-13-stateful-processing.md:
```diff
@@ -293,8 +293,15 @@ new DoFn<KV<MyKey, MyValue>, KV<Integer, KV<MyKey, MyValue>>>() {
-# State and timers are not yet supported in Beam's Python SDK.
-# Watch this space!
+class IndexAssigningStatefulDoFn(DoFn):
+  INDEX_STATE = BagStateSpec('index', VarIntCoder())
```
As per the discussion, please update this to use CombiningState. (Possibly elsewhere as well.)

robertwb commented on website/src/_posts/2017-02-13-stateful-processing.md:
```diff
@@ -499,31 +541,60 @@ new DoFn<KV<UserId, Event>, KV<UserId, Prediction>>() {
+
+    new_prediction = model.prediction(event)
+    model.add(event)
+
```
Use ModelFromEventsFn above instead of read-modify-write?

pabloem commented:
@robertwb I've addressed some of your comments. I added a public `SumCombineFn`, but then this also unblocked `combiners.Sum.PerKey`, and `combiners.Sum.Globally`. PErhaps only the CombineFn? WDYT?

pabloem commented:
ok we can review this after merging https://github.com/apache/beam/pull/8402

pabloem commented:
@robertwb ready for review

robertwb commented on sdks/python/apache_beam/transforms/userstate_test.py:
```diff
@@ -500,14 +500,13 @@ def process(self, element, ts=DoFn.TimestampParam):
       def process(self, element, state=DoFn.StateParam(INDEX_STATE)):
         unused_key, value = element
-        next_index, = list(state.read()) or [0]
+        next_index = state.read() or 0
```
The `or 0` is no longer needed.

robertwb commented on website/src/_posts/2017-02-13-stateful-processing.md:
```diff
@@ -293,8 +293,14 @@ new DoFn<KV<MyKey, MyValue>, KV<Integer, KV<MyKey, MyValue>>>() {
+
+  def process(self, element, state=DoFn.StateParam(INDEX_STATE)):
+    unused_key, value = element
+    next_index = state.read() or 0
```
Same.

kennknowles commented on sdks/python/apache_beam/transforms/trigger.py:
```diff
@@ -98,7 +94,11 @@ def with_prefix(self, prefix):
-  """StateTag pointing to an element, accumulated with a combiner."""
+  """StateTag pointing to an element, accumulated with a combiner.
+
+  The given tag must be unique for this stage.  If CombineFn is None then
```
What is a stage? This is not a concept in the Beam model. The tag must be unique for a step / transform / DoFn.

kennknowles commented on sdks/python/apache_beam/transforms/trigger.py:
```diff
@@ -98,7 +94,11 @@ def with_prefix(self, prefix):
-  """StateTag pointing to an element, accumulated with a combiner."""
+  """StateTag pointing to an element, accumulated with a combiner.
+
+  The given tag must be unique for this stage.  If CombineFn is None then
```
And signing yourself up for a lifetime of `if combine_fn is not None` doesn't seem very pleasant...

pabloem replied:
Removed that line.

kennknowles commented on website/src/_posts/2017-02-13-stateful-processing.md:
```diff
@@ -499,31 +540,59 @@ new DoFn<KV<UserId, Event>, KV<UserId, Prediction>>() {
- - You have two state cells declared, `@StateId("model")` to hold the current
-   state of the model for a user and `@StateId("previousPrediction")` to hold
+ - You have two state cells declared,
+   `@StateId("model")/StateParam(MODEL_STATE)` to hold the current
```
Move the slash out of the code font, at least. I wonder if there is another way to present both Python and Java. Perhaps a more long-form way explanation, or a separate table and drop the syntax from the paragraph.

kennknowles commented on website/src/_posts/2017-02-13-stateful-processing.md:
```diff
@@ -499,31 +540,59 @@ new DoFn<KV<UserId, Event>, KV<UserId, Prediction>>() {
- - You read the current model via `modelState.read()`. Because state is also
-   per-key-and-window, this is a model just for the UserId of the Event
-   currently being processed.
+   is as before, and in Python by using an argument placeholder.
```
You know, at this point it is starting to see like it should be documentation rather than editing the blog post. I say this because blog posts are dated and it is not great to go back and rewrite them like this. I know it gets a lot of hits, but probably we just need docs and they can receive all the hits.

mxm replied:
I agree, we shouldn't be rewriting the prose section of the blog post, but I still think that adding the Python code to the blog post has value.

mxm replied:
I believe you can add span around the Java/Python code, e.g. `<span language=py>Python code</span>` which will only show the selected language. This way the prose is not distorted with two languages here.

robertwb replied:
I agree we need documentation, but I also think it's worth updating this post (and not holding that up further). I would exclude the prose changes except the ", and in Python by using an argument placeholder" phrase which I think may not be as obvious as the others. (We could do fancy things with language-specific spans, but I would also be fine letting the prose remain, with that one addition, as it was.)

mxm replied:
+1 I also think it is worth getting this in as a reference point before we add further documentation.

kennknowles commented on website/src/_posts/2017-02-13-stateful-processing.md:
```diff
@@ -278,7 +278,7 @@ new DoFn<KV<MyKey, MyValue>, KV<Integer, KV<MyKey, MyValue>>>() {
   // A state cell holding a single Integer per key+window
   @StateId("index")
-  private final StateSpec<ValueState<Integer>> indexSpec = 
+  private final StateSpec<ValueState<Integer>> indexSpec =
```
The first change to the file should be a bold and loud header about the blog post being updated.

pabloem replied:
Done.

robertwb commented on sdks/python/apache_beam/transforms/trigger.py:
```diff
@@ -78,10 +77,7 @@ class AccumulationMode(object):
-  all elements will be returned as a list, otherwise the given CombineFn
-  will be applied (possibly incrementally and eagerly) when adding elements.
-  """
+  The given tag must be unique for this stage."""
```
s/stage/step/

pabloem replied:
Done.

robertwb commented on website/src/_posts/2017-02-13-stateful-processing.md:
```diff
@@ -499,31 +543,59 @@ new DoFn<KV<UserId, Event>, KV<UserId, Prediction>>() {
-   per-key-and-window, this is a model just for the UserId of the Event
-   currently being processed.
+   is as before, and in Python by using an argument placeholder.
+ - You read the current model via `modelState.read()` / `model_state.read()`.
```
I would omit the python version of the code in the prose. I think the mapping is clear enough, especially with the disclaimer above.

pabloem replied:
I think it does no harm. Is it okay if I leave it?

robertwb replied:
It harms the readability. Having both variations is not worth stuttering of repeating tiny variations in the code. (As Max says, we can use a span with langauge classes, or I would be OK with just leaving the Java version in the prose as it was originally written.)

pabloem commented:
what should we do about this?

mxm commented on website/src/_posts/2017-02-13-stateful-processing.md:
```diff
@@ -499,31 +543,59 @@ new DoFn<KV<UserId, Event>, KV<UserId, Prediction>>() {
+   save the prediction using `previousPredictionState.write(...)`, or
+   `previous_pred_state.clear()` and `previous_pred_state.add(...)` in Python.
+   Here the decision is relative to the prior prediction output, not the last
+   one computed - realistically you might have some complex conditions here.
```
I think the only debatable change is this prose which is hard to read because it intermingles Java and Python. You can conditionally display text with a `span` element which has `class=languagepy` or `class=languagejava`.

Other than that, I'm absolutely for getting this in.

mxm commented:
@kennknowles Are you ok with this now that the prose changes have been reverted?
